### ISSUE TITLE
f-bounds and wildcards don't mix?

### ISSUE THREAD

**Issue by adriaanm** [lampepfl/dotty#6146]
I think the following should compile (it does in 2.13), but it produces the following errors (note the `LazyRef` in the error message):

```scala
  trait BS[T, S <: BS[T, S]]
  trait IS extends BS[Int, IS]

  sealed trait BSElem[T, S <: BS[_, S]] // we cannot make this S <: BS[T, S], as the `_` could be something else from T (that's what the `BSElem` implicit is for)

  /*
-- [E057] Type Mismatch Error: /Users/adriaan/git/scala/src/library/scala/collection/convert/StreamExtensions.scala:34:36 
34 |  sealed trait BSElem[T, S <: BS[_, S]]
   |                                    ^
   |Type argument S does not conform to upper bound StreamExtensions.this.BS[Any, LazyRef(S)] 
*/

  object BSElem {
    implicit val intStreamShape: BSElem[Int, IS] = ???
  }
  class Ops[A] {
    def asJavaSeqStream[S <: BS[_, S]](implicit s: BSElem[A, S]): S = ???
    /*
  -- [E057] Type Mismatch Error: /Users/adriaan/git/scala/src/library/scala/collection/convert/StreamExtensions.scala:39:35 
  39 |    def asJavaSeqStream[S <: BS[_, S]](implicit s: BSElem[A, S]): S = ???
     |                                   ^
     |Type argument S does not conform to upper bound StreamExtensions.this.BS[Any, LazyRef(S)] 
    */
  }
```

**Comment by adriaanm** [comment]
This is blocking https://github.com/scala/scala/pull/7458

**Comment by smarter** [comment]
@adriaanm @lrytz Any reason why the following wouldn't work:
```scala
trait BS[T, S <: BS[T, S]]
trait IS extends BS[Int, IS]

sealed trait BSElem[T, S <: BS[T, S]]

object BSElem {
  implicit val intStreamShape: BSElem[Int, IS] = ???
}
class Ops[A] {
  def asJavaSeqStream[S <: BS[A, S]](implicit s: BSElem[A, S]): S = ???
}
```

**Comment by adriaanm** [comment]
Sadly, no — I tried that. It breaks down when the implicit is used for
computing element types of streams with wider numeric types (byte stream
for example)

**Comment by lrytz** [comment]
Yes, I tried that as well...

**Comment by smarter** [comment]
A full example would be helpful

**Comment by lrytz** [comment]
Here's a full example that, I think, reflects the situation. The source of the problem might be because of the use of `Integer` (BS/AS/IS are defined in Java in the original). The version below compiles with scalac, but not dotty. Use the two commented-out lines instead is what you suggest above, and then the implicit is not found. Maybe it's just a matter of adding more implicit instances (for Integer), I'll check later - don't have time right now anymore :-)

```
trait St[A]
trait IntSt extends St[Int]
trait AnySt[T] extends St[T]

trait StS[T, S <: St[_]]
trait StSL {
  implicit def anyStS[T]: StS[T, AnySt[T]] = ???
}
object StS extends StSL {
  implicit val intStS: StS[Int, IntSt] = ???
}

trait BS[T, S <: BS[T, S]]
trait AS[T] extends BS[T, AS[T]]
trait IS extends BS[java.lang.Integer, IS]

trait SS[T, S <: BS[_, S], Z <: St[_]]
// trait SS[T, S <: BS[T, S], Z <: St[_]]
trait SSL {
  implicit def anySS[T]: SS[T, AS[T], AnySt[T]] = ???
}
object SS extends SSL {
  implicit val iss: SS[Int, IS, IntSt] = ???
}

object T {
  implicit class ListExt[A](l: List[A]) {
    def ext[S <: BS[_, S], Z <: St[_]](implicit sh: SS[A, S, Z], sth: StS[A, Z]): S = ???
    // def ext[S <: BS[A, S], Z <: St[_]](implicit sh: SS[A, S, Z], sth: StS[A, Z]): S = ???
  }
}

object U {
  import T._
  List(1).ext: IS
}
```

**Comment by odersky** [comment]
It turns out to be ridiculously hard to check these bounds in a way which is (a) correct, (b) typechecks the examples. The proposed fix in #4167 does not look correct to me, but even that's not straightforward to show since it's equally hard to work on test  cases: they all end up becoming so complex that nobody understand what the typechecker should accept or not. Nobody wants to go near the subject. My preferred solution would be to ban F-bounds altogether and to rely on implicit evidence instead. That would be the single best thing we could do to simplify Scala's type checker. Unfortunately, that's probably not realistic.

**Comment by lrytz** [comment]
Fixed in the PR by relaxing the bound

```
trait SS[T, S <: BS[_, _], Z <: St[_]]

...

    def ext[S <: BS[_, _], Z <: St[_]](implicit sh: SS[A, S, Z], sth: StS[A, Z]): S = ???
```

I cannot answer if dotty is right to report an error in the original example in the issue description.
